pipelines/validate_run.py

```
@@ -412,11 +412,105 @@ def check_thalamic(obj, where):
+    if raw_goals["goal"] is None and (
+        "chosen.goal" not in normalized or "rejected.goal" not in normalized
+    ):
+        errors.append(
+            f"{where}: preference needs both side goals when no top-level goal is present"
+        )
+    if len(set(normalized.values())) > 1:
+        errors.append(f"{where}: top-level and side goals must describe the same problem")
+    return errors
+
+
+def _require_reward(obj, where):
```

Ruff (0.16.1)

[warning] Missing return type annotation for private function `_require_reward`

(ANN202)